streams/src/test/java/org/apache/kafka/streams/smoketest/SmokeTestDriver.java

```diff
@@ -220,33 +230,34 @@ public static void verify(String kafka, Map<String, Set<Integer>> allData, int m
             keys.add(key);
             received.put(key, new HashSet<Integer>());
         }
-
-        int retryCount = 0;
-        int maxRetry = 360; // max three minutes (500ms * 360) (before we reach the end of records)
-
-        while (true) {
+        int retry = 0;
+        final long start = System.currentTimeMillis();
+        while (System.currentTimeMillis() - start < TimeUnit.MINUTES.toMillis(3)) {
             ConsumerRecords<byte[], byte[]> records = consumer.poll(500);
-            if (records.isEmpty()) {
-                retryCount++;
-                if (retryCount > maxRetry) break;
+            if (records.isEmpty() && recordsProcessed >= recordsGenerated) {
+                if (verifyMin(min, allData, false)
+                    && verifyMax(max, allData, false)
+                    && verifyDif(dif, allData, false)
+                    && verifySum(sum, allData, false)
+                    && verifyCnt(cnt, allData, false)
+                    && verifyAvg(avg, allData, false)
+                    && verifyTAgg(tagg, allData, false)) {
+                    break;
+                }
+                if (retry++ > MAX_RECORD_EMPTY_RETRIES) {
```

Comment:
Are we guarding against duplicated messages here? Otherwise since `recordsProcessed >= recordsGenerated` it seems meaningless to still wait for more records to come.